keps/sig-node/5328-node-declared-features/README.md

```diff
@@ -0,0 +1,995 @@
+}
+
+
+// Node object remains unchanged in spec, only status is modified.
+type Node struct {
+    ...
+    Status NodeStatus `json:"status,omitempty"`
+}
+
+```
+
+**Note:**
+
+*   Any new feature being introduced in `node.status.declaredFeatures` is considered a formal API change and must go through the API review process. This governance will be enforced by protecting the list of declared features with `api-approvers` in the OWNERS file.
+*   We currently have [Node Features](https://github.com/kubernetes/api/blob/e8d4d542f6a9a16a694bfc8e3b8cd1557eecfc9d/core/v1/types.go#L6279) and [Node Runtime Features](https://github.com/kubernetes/api/blob/e8d4d542f6a9a16a694bfc8e3b8cd1557eecfc9d/core/v1/types.go#L6251) which publish some runtime features through Node Status. They are too narrowly scoped and currently not used for scheduling pods.
+
+
+#### Declared Feature Semantics
+
+1.  Combine Interdependent Settings
+    *   If multiple settings are required to enable a feature, they must be collapsed into a single, logical declared feature.
+    *   This simplifies decision-making for control plane components like the scheduler, which should not need to understand multiple interdependent settings.
+
+2.  Presence of a Declared Feature
+    *   A feature should only be present in `node.status.declaredFeatures` if it is enabled and functional.
+    *   If a feature is disabled or unsupported, its corresponding identifier must not exist in `node.status.declaredFeatures`.
+    *   This approach ensures consistency and simplifies logic for the control plane, treating nodes that don't know about a feature the same as those that have it disabled.
+
+3.  Validation Rules
+    *   Strings within the `node.status.declaredFeatures` slice must follow the standard identifier format used for most Kubernetes resource names (kebab-case with 63 character limit).
+    *   The list would be sorted alphabetically and would not contain duplicate entries.
+    *   Kubelet validates each identifier and will discard any invalid or duplicate entries logging an error. The validation would also be enforced by [unit tests](#unit-tests).
+
+
+**Example:**
+
+```
+declaredFeatures:
+   - guaranteed-qos-pod-cpu-resize
+```
+
+### Kubelet Changes
+
+Kubelet has the primary responsibility of discovering the enabled features relevant for control plane during its bootstrap and updating `node.status.declaredFeatures`. It determines its full, static set of declared features **once** upon startup based on its configuration. This list is held in memory and is included as a part of the periodic `NodeStatus` update that the Kubelet sends to the API server. The Kubelet is the authoritative source of truth for its own declared features. If an external controller or user were to modify the `node.status.declaredFeatures` field on the Node object, the Kubelet's next periodic status update would automatically overwrite and revert those changes.
```

Comment:
nit:

> This list is held in memory and is included as a part of the periodic `NodeStatus` update that the Kubelet sends to the API server.

I think the Kubelet sends a json patch, so it won't actually be sent to the API server unless another component tried to modify it (in which case kubelet should overwrite it).